### DISCUSSION TITLE
Publish notifications from RecurringHostedServiceBase

### DISCUSSION THREAD

**Discussion by skttl** [umbraco/Umbraco-CMS#11272] [Features and ideas]
I would like to suggest that `RecurringHostedServiceBase` should publish notifications in different areas, and publicize the period and delay fields.

I think there should be notifications for 
- `RecurringHostedServiceScheduling`
- `RecurringHostedServiceScheduled`
- `RecurringHostedServiceRescheduling`
- `RecurringHostedServiceRescheduled`
- `RecurringHostedServiceExecuting`
- `RecurringHostedServiceExecuted`
- `RecurringHostedServiceStopping`
- `RecurringHostedServiceStopped`

I've been looking at doing a PR for this, but in my head it ends up doing a breaking change because I need a scope to publish the notifications.

Is there any other way to do it?

**What do I want that for?**
What I'm envisioning is that I can create a package that uses these notifications to collect stats about the running background tasks in a project.

I want to use the stats for a hangfire-esque dashboard with an overview of the background tasks.

So I want to collect when a task I scheduled for, how long it has been running etc. I'm thinking of storing these data in an SQL table, and then pull the data to a dashboard, where I want to be able to do stuff like "Run now", and "Cancel"

**Comment by bergmania** [comment]
Hi @skttl..

Sounds like a good idea.. The way to archive this without introducing a breaking changes is something like this:

```
        private readonly IEventAggregator _eventAggregator;

        [Obsolete("Use the ctor that inject parameters")]
        protected RecurringHostedServiceBase() : this(StaticServiceProvider.Instance.GetRequiredService<IEventAggregator>())
        {
        }

        protected RecurringHostedServiceBase(IEventAggregator eventAggregator)
        {
            _eventAggregator = eventAggregator;
        }
```

Then then you can use
`await _eventAggregator.PublishAsync(new RecurringHostedServiceScheduling(this))`

To publish events..

**Note that, we need to have `v9/dev` merged into ` v9/contrib` before the StaticServiceProvider works in hosted services.**

**Reply by skttl** [reply]
Thats great news - sounds like my first v9 PR is coming then :)
